pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer.go

```diff
@@ -63,12 +64,14 @@ func ApplyLocalityLBSetting(
 	locality *core.Locality,
 	proxyLabels map[string]string,
 	localityLB *v1alpha3.LocalityLoadBalancerSetting,
-	enableFailover bool,
+	outlierDetectionEnabled bool,
 ) {
 	if localityLB == nil || loadAssignment == nil {
 		return
 	}
 
+	enableFailover := outlierDetectionEnabled || features.SendUnhealthyEndpoints.Load()
```

Comment:
I think because of some questionable decisions we made years ago, this is fairly unsafe.

Currently, locality lb is "on" by default, if you have OD. Failover mode is probably not a reasonable default, but at least now it is scoped to only when you create OD, so if you just install Istio your traffic doesn't suddenly change anywhere.

In the future, if/when we enable SendUnhealthyEndpoints by default, that would change here.

I wonder if we should try to change the default to not enable locality lb... its hard to change behavior in Istio (backwards compat), but this one is pretty bad IMO. In our managed product we disabled it from the start.

Reply:
These changes will indeed break backward compatibility for users who have SendUnhealthyEndpoints (explicitly) enabled. Locality failover will start to be implicitly applied to all services, even without outlier detection, and besides being a potentially unexpected change in the default behavior caused by upgrading the control plane, it can lead to service degradation if user workloads are unevenly distributed across localities and some locality won't be able to handle the increase in the prioritized traffic from local clients.

I also agree that having locality balancing implicitly enabled by default and popping up as a side effect of configuring outlier detection is unintuitive and the documentation does not seem very clear on that matter, so it's reasonable to have it disabled by default instead.

But this change will as well break backward compatibility for users who may already rely on that implicit behavior. It will also affect locality distribution balancing that is currently applied regardless of outlier detection but is controlled by the same switch. (Also it seems that locality distribution balancing should consider if an endpoint is unready when calculating locality weights with SendUnhealthyEndpoints enabled but that doesn't seem the case at the moment and likely should be fixed.)

I will consider changes that can maintain reasonable backward compatibility and get back to discuss shortly.